src/python/pants/base/exception_sink.py

```diff
@@ -5,42 +5,202 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import datetime
+import faulthandler
 import logging
 import os
+import signal
 import sys
-from builtins import object
+import traceback
+from builtins import object, str
 
-from pants.util.dirutil import safe_file_dump
+from future.utils import binary_type
+
+from pants.base.exiter import Exiter
+from pants.util.dirutil import safe_mkdir, safe_open
+from pants.util.objects import Exactly, datatype
 
 
 logger = logging.getLogger(__name__)
 
 
+class LogLocation(datatype([
+    ('log_dir', Exactly(binary_type, type(None))),
+    ('pid', Exactly(int, long, type(None))),
+])):
+
+  # TODO: link the datatype default values ticket!
+  def __new__(cls, log_dir=None, pid=None):
+    if log_dir is not None:
+      log_dir = binary_type(log_dir)
+    return super(LogLocation, cls).__new__(cls, log_dir=log_dir, pid=pid)
+
+  @classmethod
+  def from_options_for_current_process(cls, options):
+    return cls(log_dir=options.pants_workdir, pid=os.getpid())
+
+
 class ExceptionSink(object):
   """A mutable singleton object representing where exceptions should be logged to."""
 
-  _destination = os.getcwd()
+  # TODO: see the bottom of this file where we call reset_log_location() and friends in order to
+  # properly setup global state.
+  # TODO: document all of these fields!
+  _log_location = None
+  # We need an exiter in order to know what to do after we log a fatal exception.
+  _exiter = None
+  # Where to log stacktraces to in a SIGUSR2 handler.
+  _interactive_output_stream = None
+
+  # NB: These file descriptors are kept under the assumption that pants won't randomly close them
+  # later -- we want the signal handler to be able to do no work (and to let faulthandler figure out
+  # signal safety).
+  _pid_specific_error_fileobj = None
+  _shared_error_fileobj = None
+
+  # Integer code to exit with on an unhandled exception.
+  UNHANDLED_EXCEPTION_EXIT_CODE = 1
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # TODO: ensure all set_* methods are idempotent!
+  @classmethod
+  def reset_log_location(cls, log_location):
+    """
+    Class state:
+    - Will leak the old file handles without closing them.
+    OS state:
+    - May create a new directory.
+    - May raise ExceptionSinkError if the directory is not writable.
+    - Will register signal handlers for fatal handlers (clobbering old values).
+    """
+
+    # TODO: check for noop on both of the arguments!
+    if log_location.log_dir is None:
+      if log_location.pid is None:
+        raise cls.ExceptionSinkError(
+          "No-op call to reset_log_location(): argument was {!r}."
+          .format(log_location))
```

Comment:
I found this message confusing. I don't think this is a noop. Maybe it should be "log_location must specify either a pid or a directory, but both were None"

Alternatively, you could change `LogLocation#__new__` so that it throws when both are None. That would make it easier to trace where the LogLocation in question came from.

Reply:
The difficulty here is that I'm overloading `LogLocation` (it used to be called `LogLocationRequest`) for both setting state and getting state, and those should have different semantics, like you describe. I think the clear answer is to split the two out (one for `exceptions_log_path` and one for `reset_log_location`).

Reply:
Done!